### ISSUE TITLE
"rclone cat" doesn't work with "--progress"

### ISSUE THREAD

**Issue by Blaok** [rclone/rclone#3743]
#### What is the problem you are having with rclone?

`rclone cat foo:bar --progress` won't work. It should write the content of `foo:bar` to stdout, yet it writes the progress reports.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.50.1-DEV
- os/arch: linux/amd64
- go version: go1.10.4
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Ubuntu 18.04

####  Which cloud storage system are you using? (eg Google Drive)

Doesn't matter

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone cat foo:bar --progress`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
DEBUG : Using config file from "xxx"
DEBUG : 5 go routines active
DEBUG : rclone: Version "v1.50.1-DEV" finishing with parameters ["rclone" "cat" "foo:bar" "-vv" "--progress"]
```

**Comment by ncw** [comment]
Are you saying that you think the output of `--progress` should be to stderr rather than stdout?

I suppose that would work and there are tools like `pv` which read on stdin and write on stdout but still display stuff on the terminal.

This would be quite easy to change but I wonder if it might break people's workflows...

Should it be configurable?

Thoughts and examples of what other unix commands do appreciated!

**Comment by Yexiaoxing** [comment]
According to GNU C (https://www.gnu.org/software/libc/manual/html_node/Standard-Streams.html), the error messages and diagnostics issued by the program should go `stderr`.

https://unix.stackexchange.com/questions/331611/do-progress-reports-logging-information-belong-on-stderr-or-stdout also gives some idea.

**Comment by ncw** [comment]
Interesting thanks @Yexiaoxing 

This is a bit of advice rclone should probably be following too

> With that in mind, I recommend writing progress counters only after checking whether the stream is interactive (e.g., with isatty()) or when explicitly enabled by a command-line option. That’s especially important for progress meters that rely on terminal update behavior to make sense, like %-complete bars.

One irritation moving the `--progress` output to stderr would be the interaction with `--log-file`.

Rclone currently redirects stderr to the log-file.  It does this so the logfile can capture panic backtraces - as far as I know you can't redirect these away from stderr in Go.  See https://stackoverflow.com/q/34772012/164234 for info (top answer is mine!).

**Comment by Blaok** [comment]
Thanks for your reply (and for this amazing project!). Yes, I think making the output of `--progress` go to stderr would be better. An alternative would be to disable progress report if stdout is redirected (not a terminal), but that would make it impossible to see the progress of `rclone cat` of a very large file (which is my use case).

About the interaction with `--log-file`, I agree it is not very compatible with `--progress` on stderr. However, I doubt if they would be used together. If someone is redirecting stderr to a logfile, I'd assume it is used in a script so the terminal output is unlikely to be monitored and I wouldn't expect `--progress` is used. If `--progress` and `--log-file` have to be used together, I'd say we need a separate process to write the progress on stderr...

**Comment by ivandeex** [comment]
Having data, log and statistics on well-known file descriptors is also important for seamless integration with automatic tools and workflows. Labeling this with `integration`.

**Comment by Sarke** [comment]
> With that in mind, I recommend writing progress counters only after checking whether the stream is interactive (e.g., with isatty()) or when explicitly enabled by a command-line option. That’s especially important for progress meters that rely on terminal update behavior to make sense, like %-complete bars.

I agree with this.  I would expect to see the progress updated to the terminal, even if I pipe stdout somewhere else.  I don't see the use-case for it to be printed every other line in an output file.

I posted this in the forum, on possible ways to solve this:
> 1. change it to use stderr (possible bc break if anyone is relying on it somehow)
> 2. provide an option to use stderr for the progress and other info.
> 3. detect when output is not going to terminal (but being piped), and only then use stderr/terminal for the additional info

**Comment by avollmerhaus** [comment]
This hit me today:
```
rclone --progress cat "$S3URL" | zstd  --decompress | tar xf -
```
